### ISSUE TITLE
RedundantModifier: violation in public c-tor of private inner class 

### ISSUE THREAD

**Issue by romani** [checkstyle/checkstyle#6813]
issue is detected at https://github.com/sevntu-checkstyle/sevntu.checkstyle/pull/742#issuecomment-498220119
documentation does not define a case that c-tor of inner classes should not have public on inner classes c-tors. It is not redundant modifier presence or absence of it do make slight difference - https://pmd.github.io/pmd-6.15.0/pmd_rules_java_bestpractices.html#accessorclassgeneration
 
```
/var/tmp $ javac Test.java

/var/tmp $ cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="RedundantModifier">
        </module>
    </module>
</module>


/var/tmp $ cat Test.java 
public class Test {
 void method(){
  Inner ic = new Inner();
 }
 private class Inner {
  public Inner(){}
 }
}

/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $java -jar checkstyle-8.21-all.jar -c config.xml Test.java 
Starting audit...
[ERROR] /var/tmp/Test.java:6:3: Redundant 'public' modifier. [RedundantModifier]
Audit done.
Checkstyle ends with 1 errors.
```

---------------

expected is no violations

**Comment by romani** [comment]
@rnveach and @pbludov , please attentively read and approve.

**Comment by rnveach** [comment]
>  Inner ic = new Inner();//Causes generation of accessor class

When I decompile the code, I see no accessor class. I could not find any examples of accessor classes on google, just methods.

> expected is no violations
> It is not redundant modifier presence 

What helps support that it is not redundant? It looks redundant to me as you can't instantiate the inner class outside of the top class.

**Comment by romani** [comment]
@rnveach , I am not sure what you are referring to. Inner c-tor is already public, why wrapper c-tor is required. 
Please extend a bit your comment, to make it a clear on what you mean.

Do you agree with issue ?

**Comment by pbludov** [comment]
>  It is not redundant modifier presence or absence of it do make slight difference

It depends on the Java version. I just checked. The code
```
public class TestClass {
 void method(){
  Inner ic = new Inner();
 }
 private class Inner {
  private Inner(){}
 }
}
```
compiled by JDK8 and decompiled back:
```
public class TestClass {

   void method() {
      new Inner(this, ($1)null);
   }

  class Inner {

     // $FF: synthetic field
     final TestClass this$0;

     private Inner(TestClass var1) {
        this.this$0 = var1;
     }

     // $FF: synthetic method
     Inner(TestClass var1, $1 var2) {
        this(var1);
     }
  }

  class $1 {}
}
```

compiled by JDK11 and decompiled back:
```
public class TestClass {

   void method() {
      new Inner(this);
   }

  class Inner {

     // $FF: synthetic field
     final TestClass this$0;

     private Inner(TestClass var1) {
        this.this$0 = var1;
     }
  }
}
```
Since https://openjdk.java.net/jeps/181 the compiler no more generates an accessor.

For the `public` inner c-tor both javac8 and javac11 generates code without accessors:
```
public class TestClass {
 void method(){
  Inner ic = new Inner();
 }
 private class Inner {
  public Inner(){}
 }
}
```
becomes
```
public class TestClass {

   void method() {
      new Inner(this);
   }

  class Inner {

     // $FF: synthetic field
     final TestClass this$0;

     private Inner(TestClass var1) {
        this.this$0 = var1;
     }
  }
}
```

For my opinion it is a bug in the PMD.

**Comment by romani** [comment]
> For the public inner c-tor both javac8 and javac11 generates code without accessors:

but it does not mean that modifier is redundant.
If it is redundant, it meant code functioning the same with and without "public" modifier. Please recheck this.
In your example it looks like "private" modifier on inner class is redundant, weird....

**Comment by pbludov** [comment]
> but it does not mean that modifier is redundant.

You are right. Once an inner type becomes `private` all access modifier to its members are redundant.

**Comment by rnveach** [comment]
My comment is similar to @pbludov specifying what the bytecode of the compiled class looks like to show what the 2nd constructor looks like.

I agree this isn't redundant. To me, redundant should mean that the compiled byte code is the same regardless of the modifier used and the access level is the same. Since the byte code changes, I agree it isn't redundant.

**Comment by romani** [comment]
>  redundant should mean that the compiled byte code is the same regardless of the modifier used and the access level is the same. Since the byte code changes, I agree it isn't redundant.

I absolutely agree. I think that issue should be approved. That is why I created it. Looks like you also approving it.
@pbludov , please read description of issue and clearly state "approved" or not. If not please try to explain again your concerns.

**Comment by pbludov** [comment]
This violation is absolutely redundant for java11+. It does not change the visibility. It does not affect the bytecode.
This violation is redundant for java8. It does not change the visibility. But the compiler generates different bytecode depending on this modifier. 

This means that at least we need a parameter to tune this check.

**Comment by romani** [comment]
@pbludov ,

> This violation is absolutely redundant

violation can be either correct or incorrect. 

>  for java11+. It does not change the visibility. It does not affect the bytecode.

so this mean it is redundant to be specified in code. So validation of checkstyle is correct. Issue should be closed.

>This violation is redundant for java8. It does not change the visibility. But the compiler generates different bytecode depending on this modifier.
May be we need a parameter to tune this check.

This mean it is not redundant. But jdk8 is EOL, no need to support old java. User can suppress violation, if he still use old java. No changes is required to Check.

======

@rnveach  and @pbludov ,
Are you ok to close this issue ?

**Comment by pbludov** [comment]
I'm OK to close.

**Comment by rnveach** [comment]
I am ok to close but before that, does the JLS also specify it is redundant?
Byte code may not change, but that doesn't mean access level is the same.

**Comment by romani** [comment]
i do not see it 
https://docs.oracle.com/javase/specs/jls/se8/jls8.pdf - "redundant" is only about members of interfaces.
the same for https://docs.oracle.com/javase/specs/jls/se12/jls12.pdf .

It is probably too tiny nuance. They do not want to clearly declare them as redundant. Probably they differentiate redundant and inefficient.

Lets think about this a bit later they interest to this will grow from users side.
